chalk-integration/src/lowering.rs

```diff
@@ -307,395 +278,404 @@ impl<'k> Env<'k> {
-                }
-                Item::OpaqueTyDefn(defn) => {
-                    let type_kind = defn.lower_type_kind()?;
-                    let id = OpaqueTyId(raw_id);
-                    opaque_ty_ids.insert(defn.identifier.str.clone(), id);
-                    opaque_ty_kinds.insert(id, type_kind);
+            Item::Impl(d) => {
+                for atv in &d.assoc_ty_values {
+                    let atv_id = AssociatedTyValueId(next_item_id());
+                    associated_ty_value_ids.insert((ImplId(raw_id), atv.name.str.clone()), atv_id);
                 }
-                Item::Impl(_) => continue,
-                Item::Clause(_) => continue,
-                Item::Foreign(_) => continue,
-            };
+            }
+
+            _ => {}
         }
+    }
 
-        let mut adt_data = BTreeMap::new();
-        let mut adt_reprs = BTreeMap::new();
-        let mut fn_def_data = BTreeMap::new();
-        let mut closure_inputs_and_output = BTreeMap::new();
-        let mut closure_closure_kind = BTreeMap::new();
-        let mut closure_upvars = BTreeMap::new();
-        let mut trait_data = BTreeMap::new();
-        let mut well_known_traits = BTreeMap::new();
-        let mut impl_data = BTreeMap::new();
-        let mut associated_ty_data = BTreeMap::new();
-        let mut associated_ty_values = BTreeMap::new();
-        let mut opaque_ty_data = BTreeMap::new();
-        let mut hidden_opaque_types = BTreeMap::new();
-        let mut custom_clauses = Vec::new();
-        let mut foreign_ty_ids = BTreeMap::new();
-
-        for (item, &raw_id) in self.items.iter().zip(&raw_ids) {
-            let empty_env = Env {
-                adt_ids: &adt_ids,
-                adt_kinds: &adt_kinds,
-                fn_def_ids: &fn_def_ids,
-                fn_def_kinds: &fn_def_kinds,
-                fn_def_abis: &fn_def_abis,
-                closure_ids: &closure_ids,
-                closure_kinds: &closure_kinds,
-                trait_ids: &trait_ids,
-                trait_kinds: &trait_kinds,
-                opaque_ty_ids: &opaque_ty_ids,
-                opaque_ty_kinds: &opaque_ty_kinds,
-                associated_ty_lookups: &associated_ty_lookups,
-                parameter_map: BTreeMap::new(),
-                auto_traits: &auto_traits,
-                foreign_ty_ids: &foreign_ty_ids,
-            };
+    let mut adt_ids = BTreeMap::new();
+    let mut fn_def_ids = BTreeMap::new();
+    let mut closure_ids = BTreeMap::new();
+    let mut trait_ids = BTreeMap::new();
+    let mut auto_traits = BTreeMap::new();
+    let mut opaque_ty_ids = BTreeMap::new();
+    let mut adt_kinds = BTreeMap::new();
+    let mut fn_def_kinds = BTreeMap::new();
+    let mut fn_def_abis = BTreeMap::new();
+    let mut closure_kinds = BTreeMap::new();
+    let mut trait_kinds = BTreeMap::new();
+    let mut opaque_ty_kinds = BTreeMap::new();
+    let mut object_safe_traits = HashSet::new();
+
+    macro_rules! lower_type_kind {
```

Comment:
I don't know if I prefer *this*, or using a trait where the impls are generated by a macro.

Reply:
I think I might prefer the latter. Since there's less cognitive complexity *here* (you still would have `defn.lower_type_kind()`. And the "impls" would be like `lower_type_kind!(AdtDefn, Adt)`. (You could make `LowerTypeKind` have a `Self: LowerParameterMap` bound I think so you can do `self.all_parameters()` in the default function.

Reply:
Well, actually no for `TraitDefn` and `OpaqueTyDefn`. But you should still be able to do some macro trickery.